### PR TITLE
Upgrade ABS client Go module to `azure-sdk-for-go/sdk/storage/azblob`

### LINKED ISSUE #745: [Enhancement] Switch the Azure Blob Storage client from the now deprecated `azure-storage-blob-go` to `azure-sdk-for-go/sdk/storage/azblob`
**Enhancement (What you would like to be added):**

The currently used Azure Blob Storage client go module `github.com/Azure/azure-storage-blob-go` at `v0.8.0` is extremely old and had been deprecated a while ago.

This dependency should be removed, and a maintained equivalent of this project, `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob` should be used for interfacing with the Azure Blob Storage API.

**Motivation (Why is this needed?):**

It is long overdue to remove the dependency on the deprecated module and make use of a more up to date version of the client.

Moving to the newer client go module would enable in making use of certain features ABS offers, which are planned to be used in an upcoming release of etcd-backup-restore.

### PR DESCRIPTION
**Pull request by renormalize**
**What this PR does / why we need it**:

Upgrades the client Go modules for ABS to `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob` from the deprecated `github.com/Azure/azure-storage-blob-go`

**Which issue(s) this PR fixes**:
Fixes #745 

**Release note**:

```improvement developer
Azure Blob Storage client Go module upgraded to github.com/Azure/azure-sdk-for-go/sdk/storage/azblob from the deprecated github.com/Azure/azure-storage-blob-go
```

### CODE REVIEW

File: pkg/snapstore/abs_snapstore.go
Comment (anveshreddy18) [inline pkg/snapstore/abs_snapstore.go:55]:
```suggestion
// AzureContainerClient embeds a *azcontainer.Client (its methods are overridden in tests).
```

**Review by anveshreddy18 — CHANGES_REQUESTED** [review]

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go:52]: thanks for catching this, the old code was plain stupid lol

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go]: Addressed in 
[8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go:46]: Addressed in [8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go]: Addressed in [8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go]: Addressed in [8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go]: Addressed in [8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [8ca71f4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/8ca71f456ce873f33caa3f85ee7dd81cf2d595dd).

**Review by ishan16696 — CHANGES_REQUESTED** [review]

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [2a8a9d5](https://github.com/gardener/etcd-backup-restore/pull/751/commits/2a8a9d5bbc0657fc57edc154a68c87ff8e81bb15).

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [2a8a9d5](https://github.com/gardener/etcd-backup-restore/pull/751/commits/2a8a9d5bbc0657fc57edc154a68c87ff8e81bb15).

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [2a8a9d5](https://github.com/gardener/etcd-backup-restore/pull/751/commits/2a8a9d5bbc0657fc57edc154a68c87ff8e81bb15).

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Removed this type altogether [2a8a9d5](https://github.com/gardener/etcd-backup-restore/pull/751/commits/2a8a9d5bbc0657fc57edc154a68c87ff8e81bb15) and used a function defined in the SDK itself which does the same.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: https://github.com/gardener/etcd-backup-restore/pull/751#discussion_r1713385961

File: pkg/snapstore/abs_snapstore_test.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go]: Addressed in [2a8a9d5](https://github.com/gardener/etcd-backup-restore/pull/751/commits/2a8a9d5bbc0657fc57edc154a68c87ff8e81bb15).

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [9c20bb0](https://github.com/gardener/etcd-backup-restore/pull/751/commits/9c20bb08172b78477510d00ad5f0c9b0c678cbd7).

**Review by ishan16696 — CHANGES_REQUESTED** [review]
Overall looks good, just 1 nit pick. 
I will re-view it as you might need to rebase this PR.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Addressed in [b5ab3a4](https://github.com/gardener/etcd-backup-restore/pull/751/commits/b5ab3a4d9c8ba91c6beb56744fae7be5841bf280).

**Comment by renormalize** [comment]
Thanks for your review @ishan16696!

Will let you know once the other PRs get merged and I resolve the conflicts.

File: pkg/snapstore/abs_snapstore_test.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore_test.go:29]: do you want to mock the AzureBlockBlob Client ?
```
//go:generate mockgen -package <name> -destination=mocks.go github.com/gardener/etcd-backup-restore/pkg/snapstore/ AzureBlockBlobClientI
```

Comment (renormalize) [reply pkg/snapstore/abs_snapstore_test.go:29]: I will leave this to another PR. Will create an issue to track this later.

Comment (ishan16696) [reply pkg/snapstore/abs_snapstore_test.go:29]: can you create an issue for it and mentioned it here.

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:119]: I didn't get this comment, can you please explain where you are checking the `ABS container exists` or it is a todo ?

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:432]: I didn't get this comment what will change in `Delete op` when immutability support is added for ABS.

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:57]:
```suggestion
// AzureContainerClient embeds a azure container.Client (its methods are overridden in tests).
```

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [inline pkg/snapstore/abs_snapstore.go:49]: please add a comment describing the these function.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:119]: As can be seen in: https://github.com/gardener/etcd-backup-restore/blob/76664bff40c39f209968bde06bf711424202ed3a/pkg/snapstore/abs_snapstore.go#L204-L213
when the Snapstore struct is created from the client, a validation is performed to check if the container that is passed as configuration to etcd-backup-restore exists or not.

This was moved from `GetABSSnapstoreFromClient` to `NewABSSnapStore`. This move makes it easier for testing when `NewABSSnapStoreFromClient` is used to create a mock client.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:432]: These are options that can be used to either perform a soft delete or not, etc. Was just documenting it there in case it is to be used with immutability.

Comment (ishan16696) [reply pkg/snapstore/abs_snapstore.go:432]: ok, got it but I don't think we will ever want to do soft delete of snapshot in `Delete operation`.
Moreover, I think `soft delete` is more of functionality of versioning which azure will used in object level immutability, not in bucket level immutability.
If you want to add a comment to remember it then may be you can rephrase this little bit with proper context.

> // Delete options can be mentioned once support for immutability is added

In `Delete api` call `DeleteOptions`  can be mentioned for soft deletion if required provided versioning is enabled on container.

File: pkg/snapstore/abs_snapstore.go
Comment (ishan16696) [reply pkg/snapstore/abs_snapstore.go:119]: ok, make sense

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:432]: I guess we can omit this comment for now. If it does become necessary we won't miss this anyway.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go:119]: I will be removing this comment since it is not necessary anymore.

File: pkg/snapstore/abs_snapstore.go
Comment (renormalize) [reply pkg/snapstore/abs_snapstore.go]: Added in [f316325](https://github.com/gardener/etcd-backup-restore/pull/751/commits/f316325fc1ff8d461033998818652a835ea6f9d6).

**Review by ishan16696 — APPROVED** [review]
Thanks for all changes. LGTM!!

File: pkg/snapstore/abs_snapstore_test.go
Comment (anveshreddy18) [inline pkg/snapstore/abs_snapstore_test.go:111]:
```suggestion
// DownloadStream returns the only field that is accessed from the response, which is the io.ReadCloser to the data
```
I'm guessing you meant this right?

File: pkg/snapstore/abs_snapstore_test.go
Comment (anveshreddy18) [inline pkg/snapstore/abs_snapstore_test.go:167]:
```suggestion
	contents := bytes.NewBuffer([]byte{})
	if _, err := io.Copy(contents, body); err != nil {
		return blockblob.StageBlockResponse{}, fmt.Errorf("error while staging the block: %w", err)
	}

	decodedBytes, _ := base64.StdEncoding.DecodeString(base64BlockID)
	decoded := string(decodedBytes)
    c.mutex.Lock()
	defer c.mutex.Unlock()
	c.staging[decoded] = contents.Bytes()

```
Since the only race condition can happen is with `c.staging` map, I feel we can move the lock to below so that any goroutines which fail to copy can fail fast and also to let these copy operations & decoding to go on for the other goroutines as those can be quite some time taking steps. Wdyt?

**Review by anveshreddy18 — CHANGES_REQUESTED** [review]

**Review by anveshreddy18 — APPROVED** [review]
LGTM

**Pull request merged**
